server/src/main/scala/com/cloudera/livy/server/LivyServer.scala

```diff
@@ -135,6 +163,64 @@ class LivyServer extends Logging {
     sys.props("livy.server.serverUrl") = _serverUrl.get
   }
 
+  def runKinit(): Boolean = {
+    val keytab = livyConf.get(LAUNCH_KERBEROS_KEYTAB)
+    val principal = SecurityUtil.getServerPrincipal(livyConf.get(LAUNCH_KERBEROS_PRINCIPAL),
+      server.host)
+    require(principal != null,
+      s"Kerberos requires ${LAUNCH_KERBEROS_PRINCIPAL.key} to be provided.")
+    require(keytab != null,
+      s"Kerberos requires ${LAUNCH_KERBEROS_KEYTAB.key} to be provided.")
+    val cacheFileName = kinitCacheDir + "/krb5cc_0"
+    val commands = Seq("kinit", "-kt", keytab, "-c", cacheFileName, principal)
+    val proc = new ProcessBuilder(commands: _*).inheritIO().start()
+    proc.waitFor() match {
+      case 0 =>
+        debug("Ran kinit command successfully.")
+        livyConf.set(RSCConf.Entry.KINIT_CACHE_FILE, cacheFileName)
+        kinitFailCount = 0
+        UserGroupInformation.getCurrentUser.reloginFromTicketCache()
```

Comment:
How does `reloginFromTicketCache` know to read the ticket cache being written to the custom location? The only way I can see to override that in `UserGroupInformation.java` is to use the environment variable, and you can't change the environment at this point...